### PR TITLE
[chain-filters] fix permissions check for users with no self-service

### LINKED ISSUE #24029: Users with "No self-service" permissions on tables can't use cascading filters
**Describe the bug**
When configuring cascading filters with 2 categorical fields, if a user doesn't have self-service permissions the cascading will fail and the user will have a search box only. The crazy thing is: if you click on the filters separately, you'll see the values

**Logs**
You'll get "You don't have permissions to do that." on the backend and a 403 on the FE

**To Reproduce**
1) New->question->people. Save this question and add it to a dashboard
2) on the dashboard add 2 filters: Location connected to State column, and category, connected to Source in the card. Link them, doesn't matter which. Save it
3) go to settings->admin->people-> create a new user
4) go to settings->admin->permissions. Remove from the all users the permission to do self service on the table "People"
5) Authenticate as the newly created user- >Go to the dashboard ->Click on the first filter and see the options // Click on the second filter and also see the options
6) Click on the source filter (the one that is not linked) and select any option. Then click on the second filter, you'll get a search box and the network tab will show 403

**Expected behavior**
If you have permissions to get the filter values without the cascading, you should have the permissions to do the cascading as well, even though you have "no self service"

**Screenshots**
Dashboard with linked filters seen as an admin
image
image

Dashboard with linked filters seen as a user who doesn't have self service permissions on the people table
Now you see
image
image

Now you don't
image

**Information about your Metabase Installation:**
- Your browser and the version: Brave latest
- Your operating system: Pop-OS 22.04
- Your databases: Postgres 14.2
- Metabase version: 1.43.4
- Metabase hosting environment: Docker
- Metabase internal database: Postgres 14.2

**Severity**
P1'ish?

**Additional context**
Maybe a cousin of https://github.com/metabase/metabase/issues/12654?

### PR DESCRIPTION
**Pull request by escherize**
Fixes #24029 and #13537

## Example

When selecting filters, we do this smart thing on a dashboard: when you select filter values to narrow down results we constrain the possible selections on the next filter dropdown you click.

So if you had a table resturaunts with the relevant structure:

| *kind* | *price* | ... |
| --- | --- | --- |
| pizza | `low` | |
| bbq   | `high` | |
| steak | `high` | |
| steak | `medium` | |

with 2 filters, one for *kind* and one for *price*, And you select steak for the *kind* filter:

Then the *price* filter's dropdown should only show `high` and `medium`. Thats what I am calling the constrained version of the api call.

This behavior usually works, but the _constrained_ version of the api call fails when the user does not have a permission for self-service. The non-constrained version still works without permission for self-service. So I made a change to allow the constrained version to work without a permission for self-service.

The un-constrained call to `chain-filter` works, and has the card id bound to `*card-id*`.
The cause of this bug is that the constrained call to `chain-filter` did Not work,
because it did not have a value bound for `*card-id*`. `*card-id*` is
important because it is used to determine which sort of permissions check is used in the query
processor [here](https://github.com/metabase/metabase/blob/users-with-no-self-service-permissions-can-use-cascading-filters/src/metabase/query_processor/middleware/permissions.clj#L94). `*card-id*` being nil is what caused permissions errors.

This change adds back in `*card-id*` when it is missing, therefore using
the same permission flow as the un-constrained codepath.

---

Re-bind `*card-id*` to get a genuine permission check

- also updates comments with s/`:key`/`:param-key` to match the code

### CODE REVIEW

File: src/metabase/api/dashboard.clj [hunk 1]
```diff
@@ -592,35 +596,37 @@
...
+                 has_more_values (boolean (some true? (map :has_more_values results)))]
+             ;; results can come back as [v ...] *or* as [[orig remapped] ...]. Sort by remapped value if that's the case
+             {:values          (if (sequential? (first values))
+                                 (sort-by second values)
+                                 (sort values))
+              :has_more_values has_more_values})
+           (catch clojure.lang.ExceptionInfo e
+             (if (= (:type (u/all-ex-data e)) qp.error-type/missing-required-permissions)
+               (api/throw-403 e)
+               (throw e)))))))))
```
Comment (escherize) [inline src/metabase/api/dashboard.clj:630]: The main change is making sure that `qp.perms/*card-id*` is set to the card-id. Somehow it is missing when doing constrained chained filters.

This view of the diff is a lot easier to read:

Screen Shot 2022-08-16 at 3 00 28 PM

**Comment by noahmoss** [comment]
I'm slightly confused. Where is `card-id` being bound in the non-constrained version of the API call? Don't both versions go through `GET /api/dashboard/:id/params/:param-key/values`?

**Comment by escherize** [comment]
Where does `*card-id*` get bound for the non-constrained version of the API call? Nowhere!

The constrained api calls were just exposing non-cached values! I commented out the [cache lookup](https://github.com/metabase/metabase/blob/master/src/metabase/models/params/chain_filter.clj#L582), and discovered that after a fresh restart the non-superuser was unable to even view the first filter's dropdown options, getting a 403. The non-superuser can see them after the admin looks at them.

To fix it, I'll bind card-id for all calls to this codepath, and put in another test for it.

**Comment by escherize** [comment]
If a user can access the dashboard’s collection, they should be allowed to access chain-filter dropdowns (by running mbql queries).

- a user without the dashboard collection permission cannot see the dashboard, so cannot hit the button. 

- dashboard access is checked on the [/api/dashboard/:id/params/:param-key/values](https://github.com/metabase/metabase/blob/users-with-no-self-service-permissions-can-use-cascading-filters/src/metabase/api/dashboard.clj#L634)  and [/api/dashboard/:id/params/:param-key/search/:query](https://github.com/metabase/metabase/blob/users-with-no-self-service-permissions-can-use-cascading-filters/src/metabase/api/dashboard.clj#L649).
  - that access can be used to bypass unneeded checks in the `qp.perms/check-query-permissions*` fxn.
  
  
  - Users with nodata, but that have collection permissions are able to run queries.
  - see: [this](https://github.com/metabase/metabase/blob/users-with-no-self-service-permissions-can-use-cascading-filters/src/metabase/models/permissions.clj#L139) row in the permission explainaition [docstring]( https://github.com/metabase/metabase/blob/users-with-no-self-service-permissions-can-use-cascading-filters/src/metabase/models/permissions.clj#L133-L150): 

  ```
  | Data perms? | Coll perms? | Block? | Segmented? | Can run? |
  | ----------- | ----------- | ------ | ---------- | -------- |
  ...
  |          no |         yes |     no |         no |       ✅ |
  ...
  ```

---  

So, now we use a new dynamic var: `*internal-ui-query*`. This is used because these chain-filter dropdown-populating queries have already passed the required permission checks and do not need to be checked again in `check-query-permissions*`, which is a function that is called by many parts of the codebase. `*internal-ui-query*` is bound `true` once the `api/read-check` passes on the dashboard in question. And we know that if a user can access a dashboard, then they are allowed to run queries against it.

**Comment by noahmoss** [comment]
So if I'm understanding correctly, this change would mean that the only permission check done for field value queries is on collection access? And if you have access to the collection the dashboard is in you can de facto see all field values?

This was brought up a while ago by @camsaul [in this thread](https://metaboat.slack.com/archives/C01NT034X2B/p1650486974835959) as being somewhat intentional and the discussion seems like it was not super conclusive. @luizarakaki do you think this change is OK to make at this point in time?

**Comment by camsaul** [comment]
Does this work with sandboxing in play? If I'm sandboxed I should only see sandboxed values of a column (at least that's what we decided last time). I would assume that it works because sandboxing data perms generally count as so-called "self-service" permissions . But it would be good to be sure

**Comment by luizarakaki** [comment]
While I agree that collection access should be enough to grant field value access, we are conditioning this change to the release of [custom list of values for dashboards and template tags (parameters)](https://www.notion.so/metabase/Custom-value-list-for-dashboard-filters-and-template-tags-parameters-8ddbdfa87b9e4d1c97e77d8b8c16c4ea).

This way, creators will have the option to assign a custom list of values (to avoid potential data leakage issues).

**Comment by escherize** [comment]
I read up about sandboxing, and setup a sandbox for a user, and it seems to work with chained filters. The sandboxed user can only see their allowed column values:

### Admin sees


### Sandbox user sees

**Comment by escherize** [comment]
To be reopened with the release of [custom list of values for dashboards and template tags parameters](https://www.notion.so/metabase/Custom-value-list-for-dashboard-filters-and-template-tags-parameters-8ddbdfa87b9e4d1c97e77d8b8c16c4ea).

**Pull request closed without merging**